Add summary function

aloctavodia opened the pull request:
Closes #11 

This adds the basic functionality for the `summary` function. It is still not on par with the `summary` function from old ArviZ, but it is better than nothing. ~Currently, the output is a dataset. To get a Pandas' DataFrame we need arviz-devs/arviz-base#25~

In the old arviz we had a `kind` argument and a `stat_focus` argument. The combination of both specifies which stats/diagnostics were computed. Now we achieve the same with just the `kind` argument. For instance `kind="all"` is the equivalent of the old `kind=all` and `stats_focus="mean"` and kind  `kind="all_median"` is the equivalent of the old `kind=all` and `stats_focus="median"`. Having a single argument seems clear to me.
 
One source of inconsistency is that previously we returned the `hdi` if `stat_focus="mean"` and the `eti` if `stat_focus="median` . But in arviz 1.0 whether to compute these two different CI depend on a global `rcparam` value. ~Should we ignore the `rcparam` specification and return what the `kind` argument dictates?~ If "stats_median", or "all_median" we return `eti` irrespective of the value of `ci_kind` (or the value in rcparams).


<!-- readthedocs-preview arviz-stats start -->
----
📚 Documentation preview 📚: https://arviz-stats--51.org.readthedocs.build/en/51/

<!-- readthedocs-preview arviz-stats end -->

amaloney commented on src/arviz_stats/summary.py:
```diff
@@ -0,0 +1,222 @@
+"""Summaries for various statistics and diagnostics."""
+
+import xarray as xr
+from arviz_base import dataset_to_dataframe, extract, rcParams
+from xarray_einstats import stats
```
```suggestion
from typing import Hashbale, Literal, Sequence

import xarray as xr
from arviz_base import dataset_to_dataframe, extract, rcParams
from xarray.core.datatree import DataTree
from xarray_einstats import stats
```

amaloney commented on src/arviz_stats/summary.py:
```diff
@@ -0,0 +1,222 @@
+        If ``None`` (default), interpret var_names as the real variables names.
+        If "like", interpret var_names as substrings of the real variables names.
+        If "regex", interpret var_names as regular expressions on the real variables names.
+    group : {"prior", "likelihood"}, default "prior"
```
default in the function signature is `posterior`. I think there is a pre-commit directive we can use to ensure the function signature match the docstrings.

aloctavodia replied:
that would be nice!

amaloney commented on src/arviz_stats/summary.py:
```diff
@@ -0,0 +1,222 @@
+        ess_tail = dataset.azstats.ess(dims=sample_dims, method="tail").expand_dims(
+            summary=["ess_tail"]
+        )
+        rhat = dataset.azstats.rhat(dims=sample_dims).expand_dims(summary=["R̂"])
```
nice font!

amaloney reviewed (changes requested):
This looks great. Just a few nits about the docstring and type-hints, if you are okay with the suggested changes.

The only thing I would request an actual change for is the default value in the docstring does not match that in the function signature.

aloctavodia commented:
I fixed the docstring and added type hints. 

Please check if they are ok, feel free to make changes if you want. Not sure about the type hints for `data`, currently the function accepts DataSet, DataTree and InferenceData. Support for InferenceData should be transitional as in ArviZ 1.0 we replace it with DataTree.

amaloney reviewed (approved):
I think this is a great start. There are some new plumbings in Python 3.13 that will make more complicated types easier to handle, but I think what you have now will work just fine.
